Incorrect Velocity Unit with Reading EAGLE

stephaniehho opened the pull request:
## PR Summary

While reading the EAGLE simulation, the original yt output physical velocity (in cgs, or km/s, etc.) misses a factor of sqrt(a).
Expected: physical velocity [cm/s] = sqrt(a) * velocity_on_disk * 1e5
where 1e5 is the unit velocity in cm per second.

For details, please see the discussion in the mailing list.   I have also included the output before and after editing the code, which uses an EAGLE simulation with z=0.271. 
https://mail.python.org/mm3/archives/list/yt-users@python.org/thread/RP2FYIRCBAOYOYDS7WIATBYVH2TIK5NT/


## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Code passes flake8 checker
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

ngoldbaum commented:
One thing that is worth checking is to see whether the temperature is still correct after this change. See e.g. #1611 where @qobilidop discovered that Gadget's units for internal energy should *not* include the sqrt(a) factor.

ngoldbaum commented:
Also welcome and thanks for your first contribution to yt!

ngoldbaum commented:
It looks like the answer tests for the OWLS velocity fields failed. Not particularly surprising if the current answers are incorrect. To clear that failure you'll need to increment the OWLS answer tests by editing this part of the tests.yaml file:

https://github.com/yt-project/yt/blob/f554a60d5b73704fa99a2579ccf66ad0370dd811/tests/tests.yaml#L47-L48

changing `local_owls_001` to `local_owls_002`. That will regenerate the answers using the new values for the velocity field.

However before we do that we should check that the internal energy and temperature is still correct with this change.

stephaniehho commented:
Unfortunately, this fix in velocity gives erroneous internal energy, because the unit from yt is code_velocity**2.  Therefore with my fix above, If I ask yt to give internal energy in (cm/s)**2, it then includes an unwanted factor a (verified by dividing this output by a, which then gives the correct value).

Let u = internal energy per unit mass
In EAGLE, physical u  [(cm/s)**2] = internal_energy_on_disk * ((unit velocity in cm per second)**2),
where unit velocity in cm per second = 1e5
So, the factor a is not in this internal energy.

NOTE: With my yt-3.4.0, I don't have the unit code_specific_energy (mentioned in #1611 ) when I check the register.  With the data_structure.py on GitHub today, the exact same issues occur,  and there is no code_specific_energy in the registry.

stephaniehho commented:
Just to clarify, temperature is correct, but the internal energy is not.

brittonsmith commented:
I think you can fix this by setting `self.specific_energy_unit`  manually in `self._set_owls_eagle_units`.

ngoldbaum commented:
OK, so it sounds like you don't have the patched version you're proposing here installed. It may be helpful to read through the yt developer guide, in particular the part describing how to set up a development environment:

http://yt-project.org/docs/dev/developing/developing.html#how-to-get-the-source-code-for-editing

and how to propose changes:

http://yt-project.org/docs/dev/developing/developing.html#making-and-sharing-changes

Regardless, I think you can fix the issues with the internal energy units by adding the following to the end of the function you've modified:

```python
specific_energy_unit_cgs = self._unit_base["UnitEnergy_in_cgs"] / self._unit_base["UnitMass_in_g"]
self.specific_energy_unit = self.quan(specific_energy_unit_cgs, '(cm/s)**2')
```

ngoldbaum commented:
I'm going to go ahead and push some changes to this branch in the next couple of days so we can fix this issue. Please let me know if you would like to come back and implement the suggestion britton and I made to fix the tests.

stephaniehho commented:
Sure, I am happy to see if the velocities and specific energy get the
correct units after the new pushed fixes (using the same simulation z=0.271
that I used previously).
Or do you mean by I should clone the github version now, and add the two
lines that Britton suggested and see?


On Tue, Feb 13, 2018 at 9:19 AM, Nathan Goldbaum <notifications@github.com>
wrote:

> I'm going to go ahead and push some changes to this branch in the next
> couple of days so we can fix this issue. Please let me know if you would
> like to come back and implement the suggestion britton and I made to fix
> the tests.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yt-project/yt/pull/1675#issuecomment-365338259>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AThOLtir6ZtGYbl-YEUg4u2hceqVrhq7ks5tUcQwgaJpZM4Rr2iC>
> .
>

ngoldbaum commented:
I mean to add the lines that I suggested. You should also try building yt from source with your pull request applied, that way you'll be able to test whether the issue is fixed or not. See the instruction on how to set up a yt development environment that I linked to above.

stephaniehho commented:
I just set up the developer environment from the website you suggested (
http://yt-project.org/docs/dev/developing/developing.html#how-to-get-the-source-code-for-editing).
I changed the velocity unit to 'cm/s * sqrt(a)' (as in my pull request),
and added the two lines related to the specific energy unit that you stated.

Now, the velocity and internal energy have correct values and units, so as
the temperature (which was correct as well originally before any fixes)!

On Tue, Feb 13, 2018 at 9:54 AM, Nathan Goldbaum <notifications@github.com>
wrote:

> I mean to add the lines that I suggested. You should also try building yt
> from source with your pull request applied, that way you'll be able to test
> whether the issue is fixed or not. See the instruction on how to set up a
> yt development environment that I linked to above.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yt-project/yt/pull/1675#issuecomment-365348855>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AThOLuH5jgaUmWXhbFbtRQAZrVfL2uheks5tUcxOgaJpZM4Rr2iC>
> .
>

ngoldbaum commented:
OK, great! Now you need to commit those changes and push to this branch. You'll need to first fetch the commit from this pull request:

```
$ git fetch https://github.com/stephaniehho/yt patch-1
$ git checkout patch-1
<edit the code to fix the bug>
$ git commit -am "set specific energy unit"
$ git push https://github.com/stephaniehho/yt patch-1
```

And with that last command the pull request should be updated and we'll be able to merge this pull request. If you've never used `git` before, it may prompt you to configure a name and e-mail address before letting you commit.

ngoldbaum commented:
Nice! I think you will also need to update the OWLS answer tests since the answers we currently have stored are incorrect. You just need to edit this part of this file:

https://github.com/yt-project/yt/blob/f554a60d5b73704fa99a2579ccf66ad0370dd811/tests/tests.yaml#L47-L48

And change `local_owls_001` to `local_owls_002`. That will tell the test box to generate a new set of answers to compare with.

ngoldbaum commented:
And of course once you make that change, you'll need to commit it and then push to github so the pull request gets updated.

ngoldbaum commented:
It looks like one of the travis builds took longer than normal and timed out but otherwise the tests passed. I'm going to go ahead and merge this so I can backport it.

Thanks very much for your first contribution to yt! 🎆 💯 🥇
